superset/db_engine_specs/superset.py

```diff
@@ -0,0 +1,330 @@
+        if not self._allow_dml:
+            raise ProgrammingError(f'DML not enabled in database "{self.database}"')
+        return method(self, *args, **kwargs)
+
+    return cast(F, wrapper)
+
+
+def has_rowid(method: F) -> F:
+    @wraps(method)
+    def wrapper(self: "SupersetShillelaghAdapter", *args: Any, **kwargs: Any) -> Any:
+        # pylint: disable=protected-access
+        if not self._rowid:
+            raise ProgrammingError(
+                "Can only modify data in a table with a single, integer, primary key"
+            )
+        return method(self, *args, **kwargs)
+
+    return cast(F, wrapper)
+
+
+# pylint: disable=too-many-instance-attributes
+class SupersetShillelaghAdapter(Adapter):
+
+    """
+    A shillelagh adapter for Superset tables.
+
+    Shillelagh adapters are responsible for fetching data from a given resource,
+    allowing it to be represented as a virtual table in SQLite. This one works
+    as a proxy to Superset tables.
+    """
+
+    safe = True
+
+    type_map: Dict[Any, Type[Field]] = {
+        bool: Boolean,
+        float: Float,
+        int: Integer,
+        str: String,
+        datetime.date: Date,
+        datetime.datetime: DateTime,
+        datetime.time: Time,
+    }
+
+    @staticmethod
+    def supports(uri: str) -> bool:
+        # An URL for a table has the format superset.database[.catalog][.schema].table,
+        # eg, superset.examples.birth_names
+        parsed = urllib.parse.urlparse(uri)
+        parts = parsed.path.split(".")
+        return 3 <= len(parts) <= 5 and parts[0] == "superset"
+
+    @staticmethod
+    def parse_uri(uri: str) -> Tuple[str, Optional[str], Optional[str], str]:
+        parsed = urllib.parse.urlparse(uri)
+        parts = parsed.path.split(".")
+        if len(parts) == 3:
+            return parts[1], None, None, parts[2]
+        if len(parts) == 4:
+            return parts[1], None, parts[2], parts[3]
+        return tuple(parts[1:])  # type: ignore
+
+    def __init__(
+        self, database: str, catalog: Optional[str], schema: Optional[str], table: str,
+    ):
+        self.database = database
+        self.catalog = catalog
+        self.schema = schema
+        self.table = table
+
+        self._rowid: Optional[str] = None
+        self._allow_dml: bool = False
+        self._set_columns()
+
+    @classmethod
+    def get_field(cls, python_type: Any) -> Field:
+        class_ = cls.type_map.get(python_type, Blob)
+        return class_(filters=[Equal, Range], order=Order.ANY, exact=True)
+
+    def _set_columns(self) -> None:
+        from superset.models.core import Database
+
+        database = (
+            db.session.query(Database).filter_by(database_name=self.database).first()
+        )
+        if database is None:
+            raise ProgrammingError(f"Database not found: {self.database}")
+        self._allow_dml = database.allow_dml
+
+        # verify permissions
+        g.user = current_user
```

Comment:
Why do we need this? `g.user` is already set by FAB

Reply:
I added this early when testing as a separate package, I don't think this is needed anymore.

Reply:
Actually, @dpgaspar... I tried removing this and it fails with:

```
superset error: '_AppCtxGlobals' object has no attribute 'user'
```

Any idea why?